oauth2_provider/views/oidc.py

```diff
@@ -120,3 +147,246 @@ def _create_userinfo_response(self, request):
+
+    key = validator._get_key_for_token(token)
+    if not key:
+        return None, None
+
+    try:
+        if oauth2_settings.OIDC_RP_INITIATED_LOGOUT_ACCEPT_EXPIRED_TOKENS:
+            # Only check the following while loading the JWT
+            # - claims are dict
+            # - the Claims defined in RFC7519 if present have the correct type (string, integer, etc.)
+            # The claim contents are not validated. `exp` and `nbf` in particular are not validated.
+            check_claims = {}
+        else:
+            # Also validate the `exp` (expiration time) and `nbf` (not before) claims.
+            check_claims = None
+        jwt_token = jwt.JWT(key=key, jwt=token, check_claims=check_claims)
+        claims = json.loads(jwt_token.claims)
+
+        # Assumption: the `sub` claim and `user` property of the corresponding IDToken Object point to the
+        # same user.
+        # To verify that the IDToken was intended for the user it is therefore sufficient to check the `user`
+        # attribute on the IDToken Object later on.
+
+        return IDToken.objects.get(jti=claims["jti"]), claims
+
+    except (JWException, JWTExpired, IDToken.DoesNotExist):
+        return None, None
+
+
+def _validate_claims(request, claims):
+    """
+    Validates the claims of an IDToken for use with OIDC RP-Initiated Logout.
+    """
+    validator = oauth2_settings.OAUTH2_VALIDATOR_CLASS()
+
+    # Verification of `iss` claim is mandated by OIDC RP-Initiated Logout specs.
+    if "iss" not in claims or claims["iss"] != validator.get_oidc_issuer_endpoint(request):
+        # IDToken was not issued by this OP, or it can not be verified.
+        return False
+
+    return True
+
+
+def validate_logout_request(request, id_token_hint, client_id, post_logout_redirect_uri):
+    """
+    Validate an OIDC RP-Initiated Logout Request.
+    `(prompt_logout, (post_logout_redirect_uri, application))` is returned.
+
+    `prompt_logout` indicates whether the logout has to be confirmed by the user. This happens if the
+    specifications force a confirmation, or it is enabled by `OIDC_RP_INITIATED_LOGOUT_ALWAYS_PROMPT`.
+    `post_logout_redirect_uri` is the validated URI where the User should be redirected to after the
+    logout. Can be None. None will redirect to "/" of this app. If it is set `application` will also
+    be set to the Application that is requesting the logout.
+
+    The `id_token_hint` will be validated if given. If both `client_id` and `id_token_hint` are given they
+    will be validated against each other.
+    """
+
+    id_token = None
+    must_prompt_logout = True
+    if id_token_hint:
+        # Only basic validation has been done on the IDToken at this point.
+        id_token, claims = _load_id_token(id_token_hint)
+
+        if not id_token or not _validate_claims(request, claims):
+            raise InvalidIDTokenError()
+
+        if id_token.user == request.user:
```

Comment:
1. This case is not planed for by the specifications. We also added this functionality in our own deployment. If this is added, it should probably be a separate PR.
> An RP requests that the OP log out the End-User by redirecting the End-User's User Agent to the OP's Logout Endpoint.

2. This is the intended behavior from my reading of the specifications. 
> Furthermore, the OP MUST ask the End-User this question [asking to confirm the logout] if an id_token_hint was not provided or if the supplied ID Token does not belong to the current OP session with the RP and/or currently logged in End-User.

My interpretation is that the user should be asked for confirmation if any of the conditions is not fulfilled. If a user does not have a session then the session does not match with the ID Token. But I am also no expert in reading and understanding specifications. It might be worthwhile to have a look at the behavior of already existing implementations (e.g. Auth0, Zitadel)

Reply:
English isn't my native language, that might be why I didn't understand the specs the same way you did.

1) I'm unclear if "End-User's User Agent" also matches the RP server.
On [Wikipedia](https://en.wikipedia.org/wiki/User_agent) we can see :
> In computing, a user agent is any software, acting on behalf of a user, which "retrieves, renders and facilitates end-user interaction with Web content".
Isn't the RP sever performing an action on behalf of the user ?
If not, I agree on the separated PR.

2) You are right. If the django session is expired, then we need to prompt the user before removing the tokens.
[Keycloak does work this way](https://www.keycloak.org/docs/latest/securing_apps/index.html#logout) (supplying the id_token_hint is enougth to skip confirmation)  but maybe the case doesn't exists in Keycloak because both (session and id token) have the same lifespan.

Reply:
@tonial  the spec specifically refers to https://www.rfc-editor.org/rfc/rfc7230.html for the definition of user agent.  In this case "End User Agent", is referring to something use directly by the end user, where a server relaying a request for a user would be a intermediate user agent or proxy.   The interpretation in this case should be an application directly under the user control such as a web browser or application such as a mobile app or desktop app that can act as an agent.

Reply:
Thanks for the explanation !